### DISCUSSION TITLE
Connection problem to SUN2000-10KTL-M1

### DISCUSSION THREAD

**Discussion by ralf766** [wlcrs/huawei_solar#217] [Setting up the integration]
I installed HACS and Huawei Solar Integration v1.1.6 and tried to connect to my new Huawei solar inverter SUN2000-10KTL-M1 from within Home Assistant by adding the Huawei Solar integration. 
The firmware version of the inverter is V100R001C00SPC146. Additionally installed is the SDongleA-05 with firmware version V100R001C00SPC125. The IP address of the dongle is 192.168.178.59.
I checked, that port 502 is open at that IP address (MODBUS_TCP is activated on the dongle with unrestricted access) via 
nmap -p 502 192.168.178.59.
But if I try to add the Huawei Solar integration for the first time in network mode with IP address 192.168.178.59 and port 502 (slave address 0), the dialog "in progress" animation won't stop and will run forever until I close the dialog.
If I try to add the integration again, I get an error message "Failed to connect". After waiting a while, I will again get a not stopping "in progress" animation if I try to add the integration, but the dialog won't close and it doesn't seem to work.
Looking at the log file it seems that the connection to port 502 did succeed (at least the first time), but why is the "Add integration dialog" not closing and why can't I add the integration? Is there something else missing?

[home-assistant.log](https://github.com/wlcrs/huawei_solar/files/10308133/home-assistant.log)

Update: The firmware of the dongle was updated to V100R001C00SPC130, but it didn't change anything.

**Comment by alecsmro** [comment]
Hi @ralf766! Did you managed to solve the problem? I have the same issue :(

**Reply by ralf766** [reply]
No - the problem still persists. 
I will do some experiments with the Simply Modbus TCP Client (https://www.simplymodbus.ca/TCPclient.htm) to check, if this tool can connect to port 502.

**Comment by alexualbu** [comment]
I have the same problem with 5KTL - the debug logs say connected then disconnected.
Did you manage to fix it somehow?

**Reply by alexualbu** [reply]
@ralf766 , @alecsmro -  i got to the bottom of it - turns out i had to use slave_id 1, but if you start the HA add integration flow and add the wrong values I believe the client does not close properly (it continues to output debug logs). I could only get it to work after restarting home assistant (after validating it works directly in Python with the huawei-solar package)

**Reply by ralf766** [reply]
How did you try it directly in Python?
I tried to read a value with the Simply Modbus TCP Client. It could connect to port 502, but failed to read a register.
So more or less the same result as with Huawei Solar Integration.
@alexualbu: which firmware versions do you use?

**Comment by ralf766** [comment]
Hm - strange. I just tried it again using Simply Modbus TCP Client and it worked!
I could read e. g. register 30015 using slave id 1 (as @alexualbu, thank you for that).
So I tried again Huawei Solar Integration and for whatever reason it did also work! I'm pretty sure I tried slave id 1 before, so I don't know why it did not work before and does work now, but finally success!
The integration found the inverter, the battery and the power meter, but only two entities.
Any idea why there are only two entities?

**Comment by ralf766** [comment]
Ah - it took a while. Now I have 60 entities.

**Comment by TMoersch** [comment]
you need the newest firmware and have to enable Modbus-TCP ==> unrestricted or register your raspy IP in FusionSolar. 
This can be made in Installer Account.
I have the same inverter and it works.
Static IP in the dongle configuration is necesaary too, dhcp may change the inverters IP during reboot and cause trouble to connect.

**Comment by damshh** [comment]
Or make mac address statiic in router

**Comment by henfri** [comment]
Here's a script to test it manually, using the same package as the integration:
```
import asyncio
from huawei_solar import HuaweiSolarBridge

async def get_response():
    bridge = await HuaweiSolarBridge.create(host="192.168.177.178", port=502,slave_id=1)
    print(await bridge.update())

asyncio.run(get_response())
```

Do you have Modbus-RTU connected? That interfered in my case

**Comment by achtbitunix** [comment]
Hi all,

I'm able to connect to the inverter from a linux server and get data over IP from dongle on port 502. Thanks to @henfri for his very simple working code.

While trying to get the integration running, I'm stuck

I've set debug like mentiond in the Wiki
logger:
  logs:
    pymodbus: debug # only include this if you're having connectivity issues
    huawei_solar: debug
    homeassistant.components.huawei_solar: debug
 
This is all from home-assistant.log
2023-06-18 22:04:49.221 INFO (SyncWorker_8) [homeassistant.util.package] Attempting install of huawei-solar==2.2.7b5
2023-06-18 22:05:48.194 INFO (MainThread) [huawei_solar.huawei_solar] Waiting for connection ...

HA 2023.6.2 is running in a docker container on Linux.

atm I'm a bit lost why a connection to the inverter from inside the docker container is not working.
The code from @henfri is working from a python container.

Another observation, after starting the integration, even after the timeout, the inverter modbus is blocked until I restart HA.

Any help on this is appreciated

**Reply by henfri** [reply]
Glad my Code is helpful.
Maybe try the Code from inside the HA docker?

**Comment by achtbitunix** [comment]
Just installed huawei_solar inside the HA container. I know, nor really recommended as it's gone after HA restart.
The good news, I'm able to reach the inverter and get data.

The bad ones, the integration is still not working and I don't know why

**Comment by achtbitunix** [comment]
I found a solution for at least my problem.
Changing manifest.json to   "requirements": ["huawei-solar==2.2.6"], the integration was able to install huawei-solar and I am able to use it.
2023-06-19 14:07:49.451 INFO (SyncWorker_7) [homeassistant.util.package] Attempting install of huawei-solar==2.2.6

@wlcrs
